docs/architecture-decisions/2024-01-13_native_binaries.md

```diff
@@ -0,0 +1,182 @@
+
+The north star for this decision is to provide a distribution model that can be described as:
+
+```mermaid
+graph TD
+    A[GitHub Runners] -->|Windows| B[Packaging tool]
+    A -->|Mac| C[Packaging tool]
+    A -->|Linux| D[Packaging tool]
+    B --> E[Windows Binary]
+    C --> F[Mac Binary]
+    D --> G[Linux Binary]
+    E -->|Winget| H[Windows Users]
+    F -->|Brew| I[Mac Users]
+    G -->|Snap| J[Linux Users]
+    A -->|Wheel Build| K[Poetry]
+    K --> L[PyPi]
+    L -->|pipx| M[Python Users]
+```
+
+> Note that PyPi in the diagram above represents the existing distribution model for AlgoKit CLI prone to environment-specific bugs.
+
+The scope of this ADR only concerns the packaging for the CLI. The distribution via `snap`, `winget` and etc will be handled separately/in-parallel after decision and implementation of this ADR is in place.
+
+## Requirements
+
+- The native binaries should be easy to maintain and understand from a CI/CD deployment perspective.
+- The solution should support a wide variety of Linux distributions, macOS (both Apple Silicon and Intel architectures), and Windows.
+- The solution should integrate seamlessly with existing installation options, including Homebrew, or provide an easier alternative.
+- The solution should be designed with future scalability in mind, allowing for the addition of support for other variations of architectures or else as needed.
+- The solution should not significantly increase the complexity of the build process.
+- The solution should provide clear error messages and debugging information to assist in troubleshooting any issues that may arise.
+
+## Options
+
+### Option 1 - PyInstaller
+
+**Pros**
+
+- Easy to use and configure
+- Supports multiple platforms and architectures
+- Can handle complex packages and dependencies
+- Generates a single file executable
+- Active development and community support
+- Fairly fast build time via ci - ~3-4 minutes
+- Fairly small executable size (see benchmarking results below)
+- Marginally equal executable load time in `onedir` mode compared to `onefile` mode
+
+**Cons**
+
+- Occasionally requires manual configuration for more complex packages
+- Requires complex build packaging matrix to support multiple platforms and architectures
+- Requires minor tweaks in algokit cli to account for the fact that features relying on `sys.executable` will point to algokit cli executable instead of python interpreter. This is a minor change and can be done in a backwards compatible way however still a con to consider.
```

Comment:
Is there a `sys.executable` equivalent that works?

Reply:
@robdmoore we can still invoke anything within copier tasks so far i was thinking as a fix for this we can work on extending algokit cli to include something like `algokit run` which can act very similarly to `algokit deploy` and using that to wrap and invoke the inject.py properly. Plus we probably will cut dependency on inject.py execution via copier as part of work on init v2 and something like algokit run might get handy there as well.

Reply:
@achidlow @daniel-makerx let me know what you think about this as well. 

To recap, the problem as highlighted in the adr is due to the fact that when running copier on full stack from a binary, the `sys.executable` leads to the binary not the python interpeter bundled with that binary. What do you think about introducing `algokit run` to be used to wrap execution of python files (or similarly to deploy, any commands). That way we can specify something like `algokit run 'python inject.py'` in the copier tasks instead of relying on sys.executable aliased variable named `_copier_python`